src/libraries/Common/src/Interop/Unix/System.Native/Interop.ReadLink.cs

```diff
@@ -21,40 +21,58 @@ internal static partial class Sys
         /// Returns the number of bytes placed into the buffer on success; bufferSize if the buffer is too small; and -1 on error.
         /// </returns>
         [LibraryImport(Libraries.SystemNative, EntryPoint = "SystemNative_ReadLink", SetLastError = true)]
-        private static partial int ReadLink(ref byte path, byte[] buffer, int bufferSize);
+        private static partial int ReadLink(ref byte path, ref byte buffer, int bufferSize);
 
         /// <summary>
         /// Takes a path to a symbolic link and returns the link target path.
         /// </summary>
         /// <param name="path">The path to the symlink.</param>
         /// <returns>Returns the link to the target path on success; and null otherwise.</returns>
-        internal static string? ReadLink(ReadOnlySpan<char> path)
+        internal static string? ReadLink(ReadOnlySpan<char> path) => ReadLink(path, false, false);
+
+        /// <summary>
+        /// Takes a path to a symbolic link and returns the link target path.
+        /// </summary>
+        /// <param name="path">The path to the symlink.</param>
+        /// <param name="isDirectory">true if the <paramref name="path"/> is known to be a directory; otherwise, false.</param>
+        /// <param name="validatePath">true if <paramref name="path"/> need to be validated; otherwise, false.</param>
+        /// <returns>Returns the link to the target path on success; and null otherwise.</returns>
+        internal static string? ReadLink(ReadOnlySpan<char> path, bool isDirectory, bool validatePath)
         {
-            int outputBufferSize = 1024;
+            int outputBufferSize = 256;
 
             // Use an initial buffer size that prevents disposing and renting
             // a second time when calling ConvertAndTerminateString.
-            using var converter = new ValueUtf8Converter(stackalloc byte[1024]);
+            using var converter = new ValueUtf8Converter(stackalloc byte[outputBufferSize]);
+            Span<byte> stackBuffer = stackalloc byte[outputBufferSize];
+            ref byte pathReference = ref MemoryMarshal.GetReference(converter.ConvertAndTerminateString(path));
+            bool bufferToResize = false;
+            string? result = InternalReadLink(ref pathReference, ref MemoryMarshal.GetReference(stackBuffer), stackBuffer.Length, ref bufferToResize, isDirectory, validatePath);
 
+            if (result != null)
+            {
+                return result;
+            }
+            else if (!bufferToResize)
+            {
+                return null;
+            }
+
+            outputBufferSize *= 2;
```

Comment:
We shouldn't need two different code paths, one to handle the first attempt and another to handle subsequent.  We typically do this instead with code along the lines of (pseudo-code):
```C#
byte[]? arrayPoolBuffer = null;
Span<byte> span = stackalloc byte[...];
try
{
    while (true)
    {
        if (DoWork(span))
        {
            ... // successful, process the data
            return result;
        }

        ... // unsuccessful, double buffer size and loop around to try again
        byte[]? oldBuffer = arrayPoolBuffer;
        arrayPoolBuffer = ArrayPool<byte>.Shared.Rent(span.Length * 2);
        span.CopyTo(arrayPoolBuffer);
        if (oldBuffer is not null)
        {
            ArrayPool<byte>.Shared.Return(oldBuffer);
        }
     }
}
finally
{
    if (arrayPoolBuffer is not null)
    {
        ArrayPool<byte>.Shared.Return(arrayPoolBuffer);
    }
}
```